### ISSUE TITLE
3.0.0 roadmap?

### ISSUE THREAD

**Issue by h-vetinari** [cython/cython#4022]
There was a steady stream of alpha releases in the first half of last year, but the last alpha was ~7 months ago. I realize the difficulty of pulling off such an ambitious refactor, yet still wanted to respectfully ask what the current status / roadmap / plan is.

**Comment by robertwb** [comment]
Cython is run by a very small set of unpaid developers, with a high variance in the amount of time they are able to put in. 3.0 is still something we'd like to do, but it's not something we can assign a timeline to.

**Comment by leofang** [comment]
I am very interested in Cython 3's roadmap/plan too, and I suppose it does not necessarily need to be tied to a committed timeline 🙂 At least to me getting to know what's ahead of us with Cython 3 is enough. A more interesting question is the roadmap (again, no timeline needed) for Cython 0.29.x *after* Cython 3.0 is out (would it be deprecated/unsupported/etc).

**Comment by auvipy** [comment]
I think version 3 development has come a long way https://github.com/cython/cython/milestone/58 , maybe after another alpha release, it will be easier to find out other blockers and figure out a possible rough timeline. also, more requent contribuotrs could be added as collaborator & fundraising could be run.

**Comment by scoder** [comment]
I'll try to get a new release out during the next days, to get things rolling again.

**Comment by h-vetinari** [comment]
Not sure if this is worth a separate issue, but wanted to ask if the final version is supposed to be `3.0` (since the alphas are called `3.0a7` etc.) or `3.0.0`?

Currently it's a three-part number (`0.29.23`), and I think it should stay that way. Presumably the project will use SemVer? If not, the intended version scheme should be noted somewhere (though, realistically, people end up expecting SemVer anyway, even when it's explicitly documented not to be, as the recent experience with `cryptography` showed...).

If it's going to be `3.0.0`, the naming of the (to-be-released) alphas should reflect that...

**Comment by scoder** [comment]
Good call @h-vetinari. We'll definitely continue to use three parts. It should be `3.0.0` then, although both would be [PEP-440](https://www.python.org/dev/peps/pep-0440/) compliant.

It's probably not strictly SemVer that we use, just like it's probably not strictly SemVer that any other project in the world uses. But increasing minor and micro versions should, on average, get better and not worse. ;-)

**Comment by h-vetinari** [comment]
I'm sorry to be "that person"... I'm just wondering if it's not worth to release all the already existing improvements in 3.0.0 sooner rather than later. I remember a time (e.g. [here](https://github.com/cython/cython/issues/2800#issuecomment-466781420)) when 3.0 had been planned to be released in 2019, which is not in any way meant as a criticism - I get how ambitious this project is, and how constrained the resources.

Still, version numbers are cheap (perhaps aside from the emotional attachment) - why not release 3.0 more or less as is, and push the blockers to 4.0? There are many improvements in the master branch that people are waiting for (or even blocked on). This would also finally open the door for removing py27, py34 & py35 support (for reference, even py36 is EOL in 3 month)...

I ask mainly because the todo-list for 3.0 has not gotten substantially shorter in recent times, and there doesn't seem to be a plan-of-attack to tackle it (which is fine!). But perhaps perfect is the enemy of good here, and pushing out a new major version could relieve some pressure and rejuvinate spirits. 🙃

**Comment by da-woods** [comment]
> version numbers are cheap

The counter argument is that there is a perfectly serviceable alpha version available so anyone interested can experience the future, now ;)

-----------

I think the main aim for 3.0 was to try to get any changes likely to break user code over in one go. I'm not sure how many of those there actually are remaining - possibly the switch to `language_level=3` and `binding=True` were the main ones. But one option would be to prioritise only those changes.

**Comment by h-vetinari** [comment]
> perfectly serviceable alpha version

I think the vast majority of downstream projects will not adopt anything but a full release. They might be testing with 3.0.0a, but they won't ever switch to depending on an alpha version. Which is one of my main points, because there's a lot of good stuff in the repo that's 3.0.0a-only at the moment.

**Comment by leofang** [comment]
Agree. No release, no switch over. It's been like this for the vast majority of projects that I know/participate. Moreover, staying in alpha would not help distro people get prepared either. The middle ground to me is to make at least an rc release.

**Comment by hmaarrfk** [comment]
one of the coolest things from 3.0 is the volatile keyword. other than installing the version from GitHub, it seems hard to use that feature.

**Comment by scoder** [comment]
@h-vetinari, I think you're right. Most of the truly critical 3.0 issues are solved. What remains with regard to incompatible changes is:
https://github.com/cython/cython/issues/4280
https://github.com/cython/cython/issues/2304

[The rest](https://github.com/cython/cython/issues?q=is%3Aopen+is%3Aissue+label%3A%22P%3A+blocker%22) are mostly bugs, which by definition aren't release critical (enough to block the 3.0 release).

Both tickets above have PRs, just not ready yet. I think that can be worked on. Then we should be close.

**Comment by da-woods** [comment]
> Both tickets above have PRs

I can't see a PR for #4280? (Although I guess the actually change would be pretty small and the work would be in the tests)

--------

@hmaarrfk 

> other than installing the version from GitHub, it seems hard to use that feature.

`pip install Cython==3.0.0a9`

**Comment by scoder** [comment]
> I can't see a PR for #4280?

Sorry, I mixed it up with another ticket. #4280 is easy to implement but has a potentially large impact.

**Comment by hmaarrfk** [comment]
Cool, i didn't know we could install alphase from pip that way. It does complicate my (conda) based environment, but maybe worth the risk.

Personally, I think that the alpha cython requirement puts me on the "too complicated" for "a half step toward pure-C implementation". But thank you for the tip.

(I might be tempted to make a conda package for myself maybe :/)

**Comment by h-vetinari** [comment]
> I might be tempted to make a conda package for myself maybe

I have a PR for conda-forge open: https://github.com/conda-forge/cython-feedstock/pull/75
The only reason it is failing is because I'm now running the full test suite, rather than just some smoke tests. There are only a handful of failures, but I'm sure it would be fine to publish those to an RC branch in the cython-feedstock. We can discuss this on the PR if you're interested.

**Comment by h-vetinari** [comment]
> @scoder: @h-vetinari, I think you're right. Most of the truly critical 3.0 issues are solved.

That's very great news!

> #4280 is easy to implement but has a potentially large impact on user code.

I think it would be fine to have some more pre-releases, but having such a "path" to 3.0.0 already is a big step forward! :)

**Comment by h-vetinari** [comment]
> @scoder [Oct 6th, 2021]: Most of the truly critical 3.0 issues are solved. What remains with regard to incompatible changes is: #4280 #2304
> 
> [The rest](https://github.com/cython/cython/issues?q=is%3Aopen+is%3Aissue+label%3A%22P%3A+blocker%22) are mostly bugs, which by definition aren't release critical (enough to block the 3.0 release).

#2304 has been solved and #4280 has a [PR](https://github.com/cython/cython/pull/4670) that's been languishing for a while - if that's the last real blocker, wouldn't it make sense to prioritize finishing this above most other feature/bugfix work? Not least because the backport of `noexcept` to 0.29 would probably need some time to percolate as well?

This would allow to do beta/rc/release of 3.0.0 over the course of the following months, and then finally permit the cython devs (& contributors) to save/recoup the time necessary for all the continual sync- & backport work that's needed for 0.29 until then.

**Comment by da-woods** [comment]
> Not least because the backport of noexcept to 0.29 would probably need some time to percolate as well?

I'll see if I can get the backport of `noexcept` done. That should be fairly easily (it's largely just copy-pasting a small bit of the PR) and I think can be done in advance of the 3.0 changes

**Comment by scoder** [comment]
The most critical change for this alpha was https://github.com/cython/cython/pull/4606. Now that that's in, I'm mostly looking through PRs that are ready to be included, then it's time for 3.0a11.

**Comment by h-vetinari** [comment]
> Now that that's in, I'm mostly looking through PRs that are ready to be included, then it's time for 3.0a11.

Great to hear!

(note I wasn't really talking about the content of the next alpha here, more about 3.0.0 overall)

**Comment by scoder** [comment]
3.0.0a11 is released. I could live with starting the beta release phase now.

Since this was the first 3.0 release to support the upcoming Py3.11, I postponed a couple of waiting PRs in order to reduce the risk of breakage for those who want to test their code with that. I'll see that I get them merged during the next days, wait a bit for feedback on this release, and then it's probably time to move on into beta.

If you have a PR hanging around that you think is ready to be included, please add a comment there to nominate it for the beta release, referring back to this issue so that they show up here.

**Comment by 0dminnimda** [comment]
Are all of those https://github.com/cython/cython/milestone/58 should be merged / implemented or will some be moved to 3.1?
It's just that some of those require new features.

**Comment by da-woods** [comment]
> Are all of those https://github.com/cython/cython/milestone/58 should be merged / implemented or will some be moved to 3.1? It's just that some of those require new features.

I suspect at least some of them will be left for later.

**Comment by maxbachmann** [comment]
Is there anything except https://github.com/cython/cython/issues/4280 that is still needed for a final Cython 3 release?

**Comment by 0dminnimda** [comment]
Probably all of the blockers from 3.0 milestones, e.g. #2912, #1274, #1434, #2079, #3651, #4193, #4280, #4936, here's a [list](https://github.com/cython/cython/issues?q=is%3Aopen+is%3Aissue+label%3A%22P%3A+blocker%22)

**Comment by h-vetinari** [comment]
> Probably all of the blockers from 3.0 milestones [...]

@scoder [mentioned](https://github.com/cython/cython/issues/4022#issuecomment-936963181) quite a while ago that the only _real_ blockers for 3.0 (notwithstanding the labels) are:
* https://github.com/cython/cython/issues/2304 (solved by now)
* https://github.com/cython/cython/issues/4280 ([PR](https://github.com/cython/cython/pull/4670) in pretty advanced state)

**Comment by h-vetinari** [comment]
> @scoder [Oct 6th, 2021]: @h-vetinari, I think you're right. Most of the truly critical 3.0 issues are solved. What remains with regard to incompatible changes is: #4280 #2304
> 
> [The rest](https://github.com/cython/cython/issues?q=is%3Aopen+is%3Aissue+label%3A%22P%3A+blocker%22) are mostly bugs, which by definition aren't release critical (enough to block the 3.0 release).

@scoder, now that these two have been solved, what "truly critical 3.0 issues" are left, if any? I'm guessing #4936 might qualify (which has a pretty advanced [PR](https://github.com/cython/cython/pull/5016), luckily), is there anything else IYO?

**Comment by scoder** [comment]
There's https://github.com/cython/cython/issues/5088 to help with migration after changing the default exception propagation in #4280. Turns out that migration isn't easy for some code bases, so we'll have to help with that. Or just provide an option to keep the unsafe legacy behaviour.

**Comment by EwoutH** [comment]
Would it be possible to move all the issues and PRs currently in the [3.0 milestone](https://github.com/cython/cython/milestone/58) to that are not essential for 3.0 to the [3.1 milestone](https://github.com/cython/cython/milestone/65)? That would give a better overview what should be focused on short term.

Are all features that need to go into 3.0 merged? If so, can we do a beta release of 3.0? That would signal that we're nearing a release and starting to test Cython 3.0 is recommended.

**Comment by jakirkham** [comment]
FWIW we are already building Cython 3.0 alphas in [conda-forge (under a separate label)]( https://anaconda.org/conda-forge/repo/files?type=any&label=cython_dev ). There's also [some package (by opting in) already using it]( https://github.com/search?q=org%3Aconda-forge%20cython_dev&type=code ).

**Comment by jjerphan** [comment]
Hi all,

Projects like scikit-learn would greatly benefit from a stable release of Cython 3.0

Is there anything major issue left to resolve for the release of Cython 3.0?
Do you need any feedback or help in the process? :slightly_smiling_face:

**Comment by scoder** [comment]
We'll try to get https://github.com/cython/cython/pull/5016 merged and then I'll work on a new release. I think we can go for a beta release this time. All breaking changes should be merged by then so that users can consider the behaviour stable, minus bugs.

As for a roadmap, we'll see how the next release is accepted, and then we can try to get the stable release out early next year. I'd then wait one or two bug fix releases before starting the work on Cython 3.1, i.e. ripping out everything that smells like Python 2.x. As soon as we create a stable 3.0.x branch, it'll become increasingly difficult to merge things with the cleaned up 3.1 master.

**Comment by dvzrv** [comment]
Hi! I would like to add, that the use of pre-release cython versions in some projects (e.g. https://github.com/psycopg/psycopg/issues/471) is a problem for downstream distributions, as they either can not package optimized packages (e.g. psycopg-c) or have to package a pre-release cython version (which most do not do).

**Comment by h-vetinari** [comment]
> We'll try to get #5016 merged and then I'll work on a new release.

In happy anticipation of a new release, I've collected a few milestones of the 3.0.0 journey:

- 2018, Oct. 27th: https://github.com/cython/cython/commit/c2de8efb67f80bff59975641aac387d652324e4e (first commit for 3.0.0, branched between [0.29](https://github.com/cython/cython/releases/tag/0.29) & [0.29.1](https://github.com/cython/cython/releases/tag/0.29.1)) ✨
- 2019, Oct. 14th: Python 3.8 released (tested since [before](https://github.com/cython/cython/commit/2d89fddbaa553af98041644f3d77983c4dbf2b09) 3.0.0 branch) 🐍 
- 2020, Feb. 11th: https://github.com/cython/cython/commit/f70d96528757d8ec6ad3ebd610cb1c6d77d4881d starts testing for Python 3.9-dev 🐍 
- 2020, Apr. 12th: [3.0a1](https://github.com/cython/cython/releases/tag/3.0a1) (1058 commits) 📦
- 2020, Apr. 23rd: [3.0a2](https://github.com/cython/cython/releases/tag/3.0a2) (84 commits) 📦
- 2020, Apr. 27th: [3.0a3](https://github.com/cython/cython/releases/tag/3.0a3) (33 commits) 📦
- 2020, May 6th: [3.0a4](https://github.com/cython/cython/releases/tag/3.0a4) (106 commits) 📦
- 2020, May 19th: [3.0a5](https://github.com/cython/cython/releases/tag/3.0a5) (113 commits) 📦
- 2020, Jul 31st: [3.0a6](https://github.com/cython/cython/releases/tag/3.0a6) (223 commits) 📦
- 2021, Feb. 28th: this issue created 📜
- 2021, Mar. 31st: https://github.com/cython/cython/commit/3df198fc0338efde39911ca1b25cc7eec7ebcda2 starts testing for Python 3.10-dev 🐍 
- 2021, May 24th: [3.0a7](https://github.com/cython/cython/releases/tag/3.0a7) (449 commits) 📦
- 2021, Jul. 2nd: [3.0.0a8](https://github.com/cython/cython/releases/tag/3.0.0a8) (79 commits) 📦
- 2021, Jul. 20th: [3.0.0a9](https://github.com/cython/cython/releases/tag/3.0.0a9) (44 commits) 📦
- 2021, Oct. 6th: last blockers [identified](https://github.com/cython/cython/issues/4022#issuecomment-936963181): [#2304](https://github.com/cython/cython/issues/2304) [#4280](https://github.com/cython/cython/issues/4280) 🛑 
- 2021, Oct. 24th: https://github.com/cython/cython/commit/9d1ffd5dcdcc0f2f719f86a9a30ce08a67641f10 starts testing for Python 3.11-dev 🐍 
- 2021, Dec. 14th: [PR](https://github.com/cython/cython/pull/4339) fixing [#2304](https://github.com/cython/cython/issues/2304) merged 🚀
- 2022, Jan. 6th: [3.0.0a10](https://github.com/cython/cython/releases/tag/3.0.0a10) (237 commits) 📦
- 2022, Jul. 30th: New blocker appears: [#4936](https://github.com/cython/cython/issues/4936) 🛑
- 2022, Jul. 30th: [3.0.0a11](https://github.com/cython/cython/releases/tag/3.0.0a11) (359 commits) 📦
- 2022, Sep. 10th: [PR](https://github.com/cython/cython/pull/4670) fixing [#4280](https://github.com/cython/cython/issues/4280) merged 🚀 
- 2022, Oct. 16th: New blocker for dealing with fall-out from [#4280](https://github.com/cython/cython/issues/4280): [#5088](https://github.com/cython/cython/issues/5088) 🛑
- 2022, Nov. 26th: As of https://github.com/cython/cython/commit/9a8609dc1c34b590838f23773d1831c2fafce2e2 Cython 3 supports 10 (🤯) CPython versions, released over a span of ~15 years 🐍
- 2022, Dec. 6th: [PR](https://github.com/cython/cython/pull/5094) fixing [#5088](https://github.com/cython/cython/issues/5088) merged 🚀
- 2022, Dec. 8th: [PR](https://github.com/cython/cython/pull/5016) fixing [#4936](https://github.com/cython/cython/issues/4936) merged 🚀 (last known / proclaimed blocker resolved 🥳)
- 2023, Jan. 25th: 243 commits and counting since 3.0.0a11, 3028 since https://github.com/cython/cython/commit/c2de8efb67f80bff59975641aac387d652324e4e
- To be ~continued~ concluded...

<details>
<summary><code>git shortlog -sn --no-merges c2de8ef..HEAD</code></summary>

```
  1551	Stefan Behnel
   343	da-woods
    45	Matus Valo
    39	Matti Picus
    35	Max Bachmann
    32	Jeroen Demeyer
    25	Zackery Spytz
    24	Robert Bradshaw
    23	mattip
    21	0dminnimda
    16	serge-sans-paille
    12	Josh Tobin
    12	Kryštof Pilnáček
    12	Thomas A Caswell
    11	Kirill Smelkov
    11	Victor Stinner
    10	jbrockmendel
     8	Ashwin Srinath
     8	Christoph Gohlke
     7	Ewout ter Hoeven
     7	jakirkham
     7	realead
     7	will-ca
     6	Lisandro Dalcin
     6	Noam Hershtig
     6	Orivej Desh
     5	Egor Dranischnikow
     5	Jonathan Helgert
     5	Matthew Edwards
     5	Nicolas Pauss
     5	Sam Sneddon
     5	Sebastian Berg
     5	samaingw
     4	Alex Henrie
     4	Alexander Shadchin
     4	Celelibi
     4	Dean Scarff
     4	Dobatymo
     4	Michael Anselmi
     4	Oleksandr Pavlyk
     4	Omer Ozarslan
     4	Pablo Galindo
     4	isotherm
     4	ptype
     4	smutch
     4	telamonian
     4	Андрей Парамонов
     3	Alex Willmer
     3	Dr Alex Meakins
     3	Elliott Sales de Andrade
     3	Felix Kohlgrüber
     3	Julien Jerphanion
     3	Marius Wachtler
     3	Michael Buesch
     3	Till Hoffmann
     3	Volker-Weissmann
     2	Alphadelta14
     2	Andrew V. Jones
     2	Antonio Valentino
     2	Arvind Natarajan
     2	Ben McGinnes
     2	Brian Wignall
     2	Brock Mendel
     2	Charles Pigott
     2	Cheuk Ting Ho
     2	Chia-Hsiang Cheng
     2	Clemens
     2	DWesl
     2	Diego Elio Pettenò
     2	Dmitry Shesterkin
     2	Eddie Elizondo
     2	Eli Schwartz
     2	Evgeni Burovski
     2	Frank Schlimbach
     2	GalaxySnail
     2	Gerion Entrup
     2	Golden Rockefeller
     2	Igor Kaplounenko
     2	Jakub Kulík
     2	Janek Bevendorff
     2	Jeppe Dakin
     2	John Kirkham
     2	Jordan Brière
     2	Kamekameha
     2	Marcel Plch
     2	Martijn van Steenbergen
     2	Matthias Braun
     2	MegaIng
     2	Mike Taves
     2	Nathan Manville
     2	Nick Coghlan
     2	Nick Pope
     2	Paul van Mulbregt
     2	Pierre Glaser
     2	Ralf Gommers
     2	Ruben Vorderman
     2	Sam Gross
     2	Tao He
     2	ThePrez
     2	Warren Weckesser
     2	William Ayd
     2	Yu Feng
     2	Yuan
     2	Yuriy Chernyshov
     2	Zhenbo Li
     2	alex-xnor
     2	bweigel
     2	cf-natali
     2	davfsa
     2	demiurg337
     2	gastineau
     2	matham
     2	mitch
     2	peng weikang
     2	pfebrer
     2	true-pasky
     2	wenjun.swj
     2	will
     2	wjsi
     2	Álvaro Abella Bascarán
     1	Alessandro Molina
     1	Alex
     1	Alexey Stepanov
     1	Andrew Brown
     1	AnyLeftovers
     1	Bluenix
     1	Chris Gyurgyik
     1	Christian Clauss
     1	Corentin Cadiou
     1	Daniele Varrazzo
     1	Elisha Hollander
     1	Emmanuel Leblond
     1	EpigeneMax
     1	Evgeny Yakimov
     1	Federico Stra
     1	Fedor Alekseev
     1	Ivan Kapeykin
     1	Jerome Kieffer
     1	Johannes Mueller
     1	Jon Dufresne
     1	Jouke Witteveen
     1	Kenrick Everett
     1	Lobstros
     1	Loisaida Sam
     1	Marcel Stimberg
     1	Mathias Laurin
     1	Matthew Brett
     1	Maximilian
     1	Maximilien Colange
     1	Michał Górny
     1	Mike Graham
     1	Molly Lloyd
     1	Naelson Douglas C. Oliveira
     1	Nguyễn Gia Phong
     1	Nicolas Hug
     1	Pasha Podolsky
     1	Pedro Marques da Luz
     1	Rajat Dash
     1	Richard Barnes
     1	Roman Donchenko
     1	Samuel Colvin
     1	Scott Wolchok
     1	Sean
     1	Sean Enck
     1	Segev Finer
     1	Shreyan Avigyan
     1	Spencer Brown
     1	Syam Gadde
     1	Søren Fuglede Jørgensen
     1	Tasha "Ren" Chin
     1	Thomas J. Fan
     1	Thomas Li
     1	Thomas VINCENT
     1	Tim Gates
     1	Tortar
     1	Vadim Pushtaev
     1	Vladimir Matveev
     1	Víctor Molina García
     1	Wenjun Si
     1	Xavier
     1	Yuri Escalianti
     1	Yury Sokov
     1	account-login
     1	aiudirog
     1	alimcmaster1
     1	aws-taylor
     1	ax487
     1	h-vetinari
     1	johnthagen
     1	luz.paz
     1	matsjoyce
     1	msg555
     1	mwtian
     1	neonene
     1	sairam4123
     1	vorfol
     1	zoj613
     1	Роман Донченко
     1	墨焓
     1	谭九鼎

```

</details>

**Comment by scoder** [comment]
3.0.0b1 is on PyPi.

Took me a while, but now it's finally there. A huge THANK YOU to everyone who contributed to it. The [list of changes](https://cython.readthedocs.io/en/latest/src/changes.html#beta-1-2023-02-25) is huge.

Windows and macOS wheels are still in the making. I'll upload them tomorrow. As always, that just means that users will get the pure Python wheel for now, which just runs a little slower but otherwise does the same thing.

Have fun trying it out. :)

**Comment by jjerphan** [comment]
Thank you, @scoder and @da-woods for all this work on Cython 3! :handshake:

**Comment by jjerphan** [comment]
Thank you, @scoder, @da-woods and all the contributors for all this work on Cython 3! :handshake:

**Comment by 0dminnimda** [comment]
I think that it would be nice to refer people to documentation page describing [Migrating from Cython 0.29 to 3.0](https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html) just in case

**Comment by maxbachmann** [comment]
Are there any plans to follow up with `3.0.0b2` in the closer future? `3.0.0b1` includes a couple of breaks, which got already fixed in main.

**Comment by da-woods** [comment]
> Are there any plans to follow up with `3.0.0b2` in the closer future? `3.0.0b1` includes a couple of breaks, which got already fixed in main.

I think I've proposed PRs for most of important breaks (although one or two aren't merged yet). The remaining one that I'd like to include but haven't yet looked at is https://github.com/cython/cython/issues/5290

So I'd be in favour of a follow up release with the main fixes relatively shortly. I don't have access to the button(s) that'd do it though.

**Comment by vyasr** [comment]
Thanks for all the fantastic work to get this far! Not to push anything, just looking to keep track of progress as per the nature of this issue: is there a roadmap or task list of some sort for what needs to happen before a non-beta 3.0 release? Are there specific bugs (or possibly features) that are viewed as blockers, or is the project planning to release betas until the maintainers are feeling confident in an official release?

**Comment by jjerphan** [comment]
Hi @vyasr, the [Cython 3.0](https://github.com/cython/cython/milestone/58) is listing remaining issues and PRs before the release of this major version.

**Comment by h-vetinari** [comment]
> Hi @vyasr, the [Cython 3.0 milestone](https://github.com/cython/cython/milestone/58) is listing remaining issues and PRs before the release of this major version.

That's not the case, and it has been addressed multiple times in this thread already.

**Comment by jjerphan** [comment]
To me, I do not know what is the source of information for the remaining items. It seems that all the major blockers that have discussed in this thread have been addressed. In the meantime, a few items have been added to the 3.0 milestone — yet, might they be postponed to Cython 3.1?

Stéfan mentioned a stable release at the beginning of this year. Even if the not all features and fixes got in, I am as a maintainer of a project using Cython really satisfied with Cython 3, and I am looking forward a stable release.

**Comment by vyasr** [comment]
> That's not the case, and it has been addressed multiple times in this thread already.

Indeed, the milestone has been treated more as a loose guideline. Totally fine if that's what the devs prefer, but was curious if there was some concrete targets they had in mind before we move from betas to an official 3.0 release.

**Comment by da-woods** [comment]
> Please speak up by the end of this week if you think we're still missing something or should include something else

From my point of view, just https://github.com/cython/cython/pull/5410 since I think it's bug people will notice.

**Comment by auvipy** [comment]
I think we can postpone python 3.12 issues for 3.0.1

**Comment by matusvalo** [comment]
>  Please speak up by the end of this week if you think we're still missing something or should include something else

Also #5478 seems to be regression in Cython 3.0.

**Comment by da-woods** [comment]
> I think we can postpone python 3.12 issues for 3.0.1

Well I think the strong hope is that Cython 3 will predate Python 3.12 so we definitely won't be able to guarantee compatibility with the final release. With that said, it's worth keeping on top of any serious issues since projects that use Cython will also be looking to test their compatibility with 3.12.

> Also https://github.com/cython/cython/issues/5478 seems to be regression in Cython 3.0.

I'll try to look into it, time permitting. But my personal judgement is that it's unlikely to affect too many people so probably isn't a "serious" regression.

**Comment by scoder** [comment]
https://github.com/actions/setup-python/issues/672 hit us during the release preparations, but I mostly got the Py2.7 CI setup working again (https://github.com/cython/cython/issues/5493). I'm in a busy week now but will try to get a release prepared soon.

I think https://github.com/cython/cython/issues/5470 is the last remaining issue in Py3.12 (no profiling support).

https://github.com/cython/cython/pull/5410 looks good now and should go into the release.

**Comment by leofang** [comment]
Q: Would it still be possible to access the docs for v0.29.x? It seems right now both version toggles (latest/stable) point to the v3.0.0 docs.

**Comment by vyasr** [comment]
Awesome work on getting this released!

**Comment by h-vetinari** [comment]
Can't believe this day arrived. 😅 

I guess it's time to close this issue now. I've updated the 3.0 [timeline](https://github.com/cython/cython/issues/4022#issuecomment-1404305257) overview, it's been quite a ride! Congratulations to the Cython team! 🥳
